Preserve Python exceptions

ketralnis opened the pull request:
Lua tends to convert its exceptions to strings, but if Python code
throws an exception we may need access to the original exception object
in order to act on the causes differently. This patch gives
LuaStateException the ability to the original object as it was raised.

ketralnis committed:
Preserve Python exceptions

Lua tends to convert its exceptions to strings, but if Python code
throws an exception we may need access to the original exception object
in order to act on the causes differently. This patch gives
LuaStateException the ability to the original object as it was raised.

ketralnis commented:
👓 @riwatt @pwildani

ketralnis commented:
I'm still not clear on whether I'd rather type:

    try:
        do_lua_stuff()
    except LuaStateException as e:
        if not isinstance(e.lua_value.to_python(), PostgresException):
            raise
        # handle it...

or

    try:
        do_lua_stuff()
    except PostgresException as e:
        # handle it...

Advantages and disadvantages to both, but what do you think?

ketralnis committed:
prevent uninitialised list

riwatt commented:
> I'm still not clear on whether I'd rather type:
> 
> ```
> try:
>     do_lua_stuff()
> except LuaStateException as e:
>     if not isinstance(e, PostgresException):
>         raise
>     # handle it...
> ```
> 
> or
> 
> ```
> try:
>     do_lua_stuff()
> except PostgresException as e:
>     # handle it...
> ```
> 
> Advantages and disadvantages to both, but what do you think?

Interesting, what do you see as pros vs cons?

pwildani commented on c/_executormodule.c:
```diff
@@ -359,24 +351,64 @@ static int translate_python_exception(lua_State *L, PyGILState_STATE gstate) {
+    lua_getfield(L, LUA_REGISTRYINDEX, EXECUTOR_LUA_CAPSULE_KEY);
+    lua_setmetatable(L, -2);
+
+    // we don't do explicit refcounting. instead we rely on our references dict
```
The old comment mentioned that we don't hold the GIL for this block. Is that still true?

ketralnis replied:
No, it's not true anymore. We now retain the GIL from both call sites

pwildani commented on lua_sandbox/executor.py:
```diff
@@ -292,8 +293,10 @@ def install_python_capsule(self):
+        # we need to be able to dealloc them. even though references is
+        # available in the control block, free_python_capsule may run *after*
+        # the control block isn't available anymore so it'll need his own copy
+        lua_pushlightuserdata(self.L, ctypes.py_object(self.references))
```
Is this the moral equivalent of an INCREF?, or is the lua stack borrowing the self.references pointer?

ketralnis replied:
The latter, this is creating the prototype for the userdatas that wrap our python objects in Lua. here we're setting self.references as an "upvalue", like a closure value so that when an instance has its __gc method called, one of the upvalues available is that references dict so he can go and delete himself from it. The weird thing about it is that Lua may unset the allocator (which contains the control block, which is usually how we'd get access to state like this) before it goes and frees everything that was allocated using that allocator. That seems... ill advised, but sadly it's what it does in real life.

So instead of just getting access to self.references via the control block like we do everywhere else, we pass it as an explicit upvalue to __gc specifically

pwildani commented:
The first implies that you're dynamically generating an exception class and instance that is is-a LuaStateException and is-a PostgresException when the exception is raised. That seems weird.
(In python3 I'd expect it to be a chained exception: `if not isinstance(e.__cause__ PostgresException): ...`)


The second implies that an internal python exception is promoted back to top level when propagating the exceptional path back through the lua stack, which seems semantically cleaner, but doesn't give an obvious way to capture the lua stack trace, which might occasionally help a lot when debugging.

pwildani commented:
Looking through the unit tests, I see `e.lua_value.to_python()`, which is semantically equivalent to `__cause__`, so I'll assume that's what you meant in the sample code, which avoids the weird is-a implications.

So the main difference is a lack of somewhere to attach the lua stack trace (if any)

ketralnis commented:
> Looking through the unit tests, I see e.lua_value.to_python(), which is semantically equivalent to __cause__

Yeah, sorry that's what I meant

> Interesting, what do you see as pros vs cons?

Pros to wrapping:

* `except LuaException` will catch anything raised in Lua, even if it had an underlying cause in the Python code
* Possible to tell whether the exception happened in Lua or in other parts
* Keep access to the Lua stack trace (which currently we drop, but not wrapping it rules this out entirely)

Pros to not wrapping:

* Looks nicer, more semantically clean
* When I port to Python3, will map more cleanly to `__cause__`

The reason I think it's worth worrying about is that it's an API change that's hard to change your mind on without being really mindful about how callers are using it. (TMK we are the only user of lua_sandbox currently but it's not like I get stats or anything)

pwildani commented:
I'm leaning towards wanting a python3-like interface, even in python2 code, mostly because python2 does not have an API convention with similar semantics. The name collision with  the compiler enforced private attribute pattern is annoying though.

* raise a LuaException from anything that kills the lua execution
* if that can be tracked to a python exception, make that accessible as the `cause`  attribute (not `__...` unless that happens to actually work) on the LuaException. (late bound or not, doesn't matter much)

ketralnis committed:
Follow Peter's advice and put the original exception on `__cause__`.

This lets us look python3-like while letting us catch everything that
halts lua execution

ketralnis commented:
went with peter's idea of setting on `__cause__` (which does in fact work), and tests and leaktest both pass

ketralnis commented:
merged to master as 22148bc
